pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Package re provides a simple function to access compile regex objects for
+// the FQDN subsystem.
+package re
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sync/atomic"
+	"unsafe"
+
+	lru "github.com/golang/groupcache/lru"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "fqdn/re")
+)
+
+// CompileRegex compiles a pattern p into a regex and returns the regex object.
+// The regex object will be cached by an LRU. If p has already been compiled
+// and cached, this function will return the cached regex object. If not
+// already cached, it will compile p into a regex object and cache it in the
+// LRU. This function will return an error if the LRU has not already been
+// initialized.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	lru := (*RegexCompileLRU)(atomic.LoadPointer(&regexCompileLRU))
+	if lru == nil {
+		return nil, errors.New("FQDN regex compilation LRU not yet initialized")
```

Comment:
Shouldn't this be a `panic` to catch misuse of this function early and loudly?

Reply:
In general we have shied away from `panic()` throughout the codebase specifically for this kind of resiliency reason. It's easy to introduce these statements with the best intentions, but then for them to slip through testing and end up causing production incidents in live environments not because of the condition that occurred, but because the agent is panicking.

The main exception is where the user has explicitly configured conflicting options, there we panic early to give the user feedback that this combination of settings is unsupported. The idea there is that as soon as the user installs / enables the option, they should get immediate feedback about the conflict.